File: AlgoExpert/Easy/palindromeCheck.py
# def isPalindrome(string):
#     reversedString = ""
#     for str in reversed(string):
#         reversedString += str
#     return reversedString == string


# def isPalindrome(string):
#     reversedChars = []
#     for str in reversed(string):
#         reversedChars.append(str)
#     return "".join(reversedChars) == string

# def isPalindrome(string, i=0):
#     j = len(string) - 1 - i
#     return True if i >= j else string[i] == string[j] and isPalindrome(string, i+1)

def isPalindrome(string):
    leftidx = 0
    rightIdx = len(string) - 1
    while leftidx < rightIdx:
        if string[leftidx] != string[rightIdx]:
            return False
        leftidx += 1
        rightIdx -= 1

    return True


print(isPalindrome("abbma"))

File: AlgoExpert/Easy/palindromeCheck.js
// Time - O(n^2), Space - O(n)
// function isPalindrome(string) {
//   let reversedString = "";
//   for (let i = string.length - 1; i >= 0; i--) {
//     reversedString += string[i];
//   }
//   return string === reversedString;
// }

// Time - O(n), Space - O(n)
// function isPalindrome(string) {
//   const reversedChars = [];
//   for (let i = string.length - 1; i >= 0; i--) {
//     reversedChars.push(string[i]);
//   }
//   return string === reversedChars.join("");
// }

//Recursion - Time - O(n), Space - O(n)
// function isPalindrome(string, i = 0) {
//   const j = string.length - 1 - i;
//   return i >= j ? true : string[i] === string[j] && isPalindrome(string, i + 1);
// }

//time - O(N), Space - O(1)
function isPalindrome(string) {
  let leftIdx = 0;
  let rightIdx = string.length - 1;
  while (leftIdx < rightIdx) {
    if (string[leftIdx] !== string[rightIdx]) return false;
    leftIdx++;
    rightIdx--;
  }

  return true;
}

console.log(isPalindrome("abcba"));
